CP-54393: VNC Console Idle Timeout

stephenchengCloud opened the pull request:
**Note that this is a draft PR for initial review**, though I've tested it with both dom0 and vm console, as:
1. This change is substantial and requires unit tests
2. Before investing time in comprehensive unit tests, I want to ensure the implementation approach is generally correct
3. Added logs for testing and will be removed

Tested:
<img width="894" height="508" alt="image" src="https://github.com/user-attachments/assets/a9dbb671-5cb7-4148-9b15-15c0509f63f1" />


Key Features
- RFB_msg_type_parser Module 
-- A simple parser: For each newly received data, only parses the message type to detect user activity 
-- Activity detection: Identifies KeyEvent (4) and PointerEvent (5) messages as user activity 
-- Unknown message handling: Skips parsing for unknown message types and continues processing

- Console_idle_monitor Module 
-- Monitors console sessions for idle timeout based on configuration, using RFB_msg_type_parser for activity detection

- Integration
-- Hooks into Unixext.proxy via a closure callback mechanism
-- No breaking changes to existing console functionalit

psafont commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -381,11 +406,38 @@ let with_polly f =
+  ; poll_event_timeout_ms= -1 (* wait indefinitely *)
+  }
+
+let proxy ?(console_init_state = default_vnc_console_state)
```
this module is not the place to put any code with knowledge of a VNC console

stephenchengCloud replied:
`proxy` needs to get the configs, at least, `is_idle_timeout_set`, because we don't want to change the behaviour when idle timeout is not set.

psafont replied:
then proxy should be elsewhere, or made generic it does not care about vnc at all so it can be kept here. VNC is not a unix concept

stephenchengCloud commented on ocaml/xapi/console.ml:
```diff
@@ -126,7 +127,309 @@ let address_of_console __context console : address option =
+      ; Unixext.poll_event_timeout_ms
+      }
+    in
+    log_in_file
```
The log is for testing and will be removed.

stephenchengCloud commented on ocaml/xapi/console.ml:
```diff
@@ -126,7 +127,309 @@ let address_of_console __context console : address option =
+  let oc = open_out_gen [Open_creat; Open_text; Open_append] 0o644 file in
+  let time_str =
+    let tm = Unix.localtime (Unix.time ()) in
+    Printf.sprintf "%04d-%02d-%02d %02d:%02d:%02d" (tm.tm_year + 1900)
```
This log will be removed like I mentioned in the commit message.

lindig commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -339,6 +339,31 @@ module CBuf = struct
     if read = 0 then x.r_closed <- true ;
     x.len <- x.len + read
+
+  (* Read data from the fd to the buffer and return the newly read data *)
```
Is this in some way re-implementing buffered IO as it is available from `in_chan`? Unbuffered IO (`Unix.read`) already reads what is available into a buffer and returns the number of bytes it has read. What problem is this code solving?

stephenchengCloud commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -339,6 +339,31 @@ module CBuf = struct
+      ( if start_pos + newly_read <= buffer_size then
+          (* No wraparound - single blit *)
+          Bytes.blit x.buffer start_pos result 0 newly_read
+        else (* Wraparound - two blits *)
```
Yes. It's much better to re-use the exsiting `read` and return buffer + offset.

stephenchengCloud commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -413,13 +465,22 @@ let proxy (a : Unix.file_descr) (b : Unix.file_descr) =
+        (fun _polly fd events () ->
           (* Do the writing before the reading *)
           if Polly.Events.(test out events) then
             if a = fd then CBuf.write b' a else CBuf.write a' b ;
```
Good idea.
We must make sure the buffer isn't modified by the checker.

stephenchengCloud replied:
After thinking about it again, this approach doesn't work. The reason is that we use the buf's `start`, `len` value in the checker to parse data that hasn't been sent to the server yet. However, after the write operation completes, it may not have written all the buffer data to the server. So when the checker checks again next time, we would be parsing the remain data from before repeatedly

minglumlu replied:
Ah. Indeed, the `write` may not get all data in buffer sent out.

minglumlu commented on ocaml/xapi/console.ml:
```diff
@@ -126,7 +127,309 @@ let address_of_console __context console : address option =
+              let has_activity =
+                List.exists
+                  (fun msg_type -> msg_type = 4 || msg_type = 5)
+                  simple_rfb_parser.received_msg_types
```
I think the logic could be as simple (two lines maybe) as:
1. refresh the state with a new`Mtime_clock.counter ()` when parsed a `KeyEvent` or `PointerEven` message.
2. after all parsing in one callback call, check the elapsed time with `Mtime_clock.count` against the idle timeout config.

minglumlu commented on ocaml/xapi/console.ml:
```diff
@@ -126,7 +127,309 @@ let address_of_console __context console : address option =
+module RFB_msg_type_parser = struct
+  type state = {
+      mutable rfb_phase: [`Handshake | `Security | `ClientInit | `Messages]
+    ; mutable incomplete_msg_remaining: int
```
It may be bytes for several messages. So it can be called as unconsumed bytes.

minglumlu commented on ocaml/xapi/console.ml:
```diff
@@ -126,7 +127,309 @@ let address_of_console __context console : address option =
+      {
+        Unixext.is_idle_timeout_set
+      ; Unixext.is_idle= false
+      ; Unixext.poll_event_timeout_ms
```
This should be in proxy rather than console/vnc.

minglumlu commented on ocaml/xapi/console.ml:
```diff
@@ -126,7 +127,309 @@ let address_of_console __context console : address option =
+          let pool = Helpers.get_pool ~__context in
+          Db.Pool.get_vm_console_idle_timeout ~__context ~self:pool
+      in
+      let poll_event_timeout_ms =
```
This should be in proxy rather than console/vnc.

minglumlu commented on ocaml/xapi/console.ml:
```diff
@@ -126,7 +127,309 @@ let address_of_console __context console : address option =
+(* https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst *)
+module RFB_msg_type_parser = struct
+  type state = {
+      mutable rfb_phase: [`Handshake | `Security | `ClientInit | `Messages]
```
Can use `variants` instead of `polymorphic variants`?
And I believe there is one phase `No_parsing` for the case that parsing failed. It's impossible to recover. The session should be exempted from the idle timeout checking.

robhoes commented:
> is idle when the amount of data flowing from the client to VNC is below a value;

Does an "idle" connection have a very predictable rate of incoming packets, e.g. request every x seconds? Could we set a threshold just above that? I guess it depends on how much traffic a non-idle but lightly used connection is expected to generate, as we don't want to cut off too eagerly, but can't set the threshold too low either

stephenchengCloud commented:
From what I observed, the data flow rate from the client depends on the applications running on the client:
- Case 1: A windows 10, runnig an application that needs to show the current time (2025-09-18 03:25:16)
<img width="279" height="249" alt="image" src="https://github.com/user-attachments/assets/ea7793a3-4ca7-4666-9709-119fc5546a53" />

Then the data from the client to the server is like this, updated every second:
Msg Type=3 is [FramebufferUpdateRequest](https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#id64) 

```
[2025-09-18 03:25:13] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:14] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:15] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:16] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:17] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:22] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:23] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:24] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:25] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:26] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:32] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
```

- Case 2: Then I closed the applications and left only an empty desktop.
I noticed there's a clock updated every minute in the corner.
<img width="180" height="41" alt="image" src="https://github.com/user-attachments/assets/137cb56a-9fef-4879-9ac6-941d57c1a303" />

Then the data from the client to the server is like this, not updated every second. But it's unpredictable.
```
[2025-09-18 03:25:54] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:25:55] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:26:00] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:26:00] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:26:46] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:26:49] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:27:01] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:27:49] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:27:54] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:28:00] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:28:52] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:29:00] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:29:00] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:29:56] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:30:01] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:30:04] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:30:58] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:31:01] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:31:09] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:32:02] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:32:14] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:33:02] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:33:04] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:33:20] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:34:02] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
[2025-09-18 03:34:25] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 4; 0; 3; 0]
```
- Case 3: A dom0 console showing an empty shell.
The data is like this, updated every 6 seconds.
```
[2025-09-18 02:50:29] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
[2025-09-18 02:50:35] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
[2025-09-18 02:50:41] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
[2025-09-18 02:50:48] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
[2025-09-18 02:50:54] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
[2025-09-18 02:51:00] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
[2025-09-18 02:51:06] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
[2025-09-18 02:51:13] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
[2025-09-18 02:51:19] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
[2025-09-18 02:51:25] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
[2025-09-18 02:51:32] [proxy] Read 10 bytes, Msg Type=3, bytes=[3; 1; 0; 0; 0; 0; 2; 128; 1; 128]
```

I believe the data rate is also rely on the network latency and dom0's workload.
Therefore, it's hard to find a proper threshold.

Let's write a simple rfb parser first. From the data I observed, the message sent from the client to the server is limited to a few types, so it shouldn't be difficult to handle.

lindig commented on ocaml/xapi/console.ml:
```diff
@@ -126,7 +127,309 @@ let address_of_console __context console : address option =
+     This prevents the parser from getting stuck on protocol variations while
+     ensuring we can still achieve our primary goal: detecting KeyEvent and
+     PointerEvent messages for activity monitoring. *)
+  let progress_handshake state data_len offset =
```
You still have to skip over the other messages. I have not looked at this in detail but decoding only some messages will only be enough if ignoring the others by knowing their length is possible without fully decoding them.

stephenchengCloud commented:
RFB parser is ready to review: https://github.com/xapi-project/xen-api/pull/6679

Close this PR for now, will reopen it after RFB parser is merged.
